[KOGITO-2758] Refactor publisher-consumer code

fjtirado opened the pull request:
Moving most of the code to constant classes. Since we cannot use reflection at runtime, Im relying on Function to execute conditional object creation. 

Result of code generation is:

For consumer

@javax.enterprise.context.ApplicationScoped()
public class TravelersMessageConsumer_3 {

    @javax.inject.Inject()
    @javax.inject.Named("Travelers")
    Process<TravelersModel> process;

    @javax.inject.Inject()
    Application application;

    @org.eclipse.microprofile.config.inject.ConfigProperty(name = "kogito.messaging.as-cloudevents")
    Optional<Boolean> useCloudEvents = Optional.of(true);

    EventConsumerFactory eventConsumerFactory = new DefaultEventConsumerFactory();

    @javax.annotation.PostConstruct()
    public void configure() {
    }

    @org.eclipse.microprofile.reactive.messaging.Incoming("travellers")
    public void consume(String payload) {
        eventConsumerFactory.get(event -> {
            TravelersModel model = new TravelersModel();
            model.setTraveller(event);
            return model;
        }, org.acme.travel.Traveller.class, TravelersMessageDataEvent_3.class, useCloudEvents).consume(application, process, payload, "travellers");
    }
}

for producer


@javax.enterprise.context.ApplicationScoped()
public class TravelersMessageProducer_7 {

    @javax.inject.Inject()
    @org.eclipse.microprofile.reactive.messaging.Channel("processedtravellers")
    org.eclipse.microprofile.reactive.messaging.Emitter<String> emitter;

    @org.eclipse.microprofile.config.inject.ConfigProperty(name = "kogito.messaging.as-cloudevents")
    Optional<Boolean> useCloudEvents = Optional.of(true);

    EventMarshaller marshaller = new DefaultEventMarshaller();

    @javax.annotation.PostConstruct()
    public void configure() {
    }

    public void produce(ProcessInstance pi, org.acme.travel.Traveller eventData) {
        emitter.send(this.marshall(pi, eventData));
    }

    private String marshall(ProcessInstance pi, org.acme.travel.Traveller eventData) {
        return marshaller.marshall(eventData, e -> new TravelersMessageDataEvent_7("", e, pi.getId(), pi.getParentProcessInstanceId(), pi.getRootProcessInstanceId(), pi.getProcessId(), pi.getRootProcessId(), String.valueOf(pi.getState()), pi.getReferenceId()), useCloudEvents);
    }
}

ricardozanini commented on api/kogito-services/src/main/java/org/kie/kogito/services/event/AbstractProcessDataEvent.java:
```diff
@@ -78,7 +107,11 @@ public String getKogitoReferenceId() {
         return this.kogitoReferenceId;
     }
 
-    public void setKogitoReferenceId(String kogitoReferenceId) {
```
Removing this won't affect JSON unmarshal/marshal in runtime?

cristianonicolai replied:
@fjtirado so why add `setKogitoStartFromNode` then? shouldnt we keep consistency?

fjtirado commented:
> Hi @fjtirado many thanks working on this, seems neat. I haven't seen a change in the CloudEvent message producer/consumer. I think it's worth taking a look and testing this change with the [Knative eventing example](https://github.com/kiegroup/kogito-examples/tree/stable/process-knative-quickstart-quarkus). You don't need a cluster to do that, a local smoke test would do.
> 
> Thanks!

I realize there is no test for consumer/producer when I see the poor coverage. Ill add one to integration test

ricardozanini commented:
Nice! But please, take your time and try to build/test the example locally as well with your changes to make sure that we are not breaking anything.

fjtirado commented:
> Nice! But please, take your time and try to build/test the example locally as well with your changes to make sure that we are not breaking anything.

I added bpmn from example to integration to ensure it is properly deployed with new code. 
Pure Unit test was also added to event implementation classes used in generated code

fjtirado commented:
> LGTM! Many thanks putting such effort to test the CloudEvent over HTTP implementation as well. I'm just curious to see if this template could be changed as well as you did here: https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-codegen/src/main/resources/class-templates/events/CloudEventsMessageProducerTemplate.java#L45-L57

Sure, done.

ricardozanini commented on kogito-codegen/src/main/resources/class-templates/events/CloudEventsMessageProducerTemplate.java:
```diff
@@ -38,33 +32,18 @@ public void produce(ProcessInstance pi, $Type$ eventData) {
     }
-
+    
+¡
```
Is this a typo? :thinking:

fjtirado commented:
jenkins retest this

fjtirado commented:
@kiegroup/gatekeepers please merge this

ricardozanini commented:
@cristianonicolai @evacchi could you please take a look?

cristianonicolai commented on integration-tests/integration-tests-quarkus/src/main/java/org/acme/travels/Traveller.java:
```diff
@@ -21,21 +21,24 @@
 	private String email;
 	private String nationality;
-	private Address address;
+    private Address address;
```
please review code formatting in this class

fjtirado replied:
I did not want to introduce noise in the review by formatting the whole class, but if it is ok, lets go ahead

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/CloudEventConsumer.java:
```diff
@@ -0,0 +1,89 @@
+    private Function<D, M> function;
+    private Class<T> cloudEventClass;
+
+    public CloudEventConsumer(Function<D, M> function,
+                              Class<T> cloudEventClass) {
```
```suggestion
    public CloudEventConsumer(Function<D, M> function, Class<T> cloudEventClass) {
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/CloudEventConsumer.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public void
+        consume(Application application, Process<M> process, String payload, String trigger) {
```
```suggestion
    public void consume(Application application, Process<M> process, String payload, String trigger) {
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/CloudEventConsumer.java:
```diff
@@ -0,0 +1,89 @@
+                                            cloudEvent.getData(),
+                                            cloudEvent.getKogitoProcessinstanceId())));
+                } else {
+                    logger
+                        .debug(
+                            "Received message without reference id, staring new process instance with trigger '{}'",
+                            trigger);
```
```suggestion
                    logger.debug("Received message without reference id, staring new process instance with trigger '{}'",
                                         trigger);
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/CloudEventConsumer.java:
```diff
@@ -0,0 +1,89 @@
+                        .debug(
+                            "Received message without reference id, staring new process instance with trigger '{}'",
+                            trigger);
+                    ProcessInstance<M> pi =
+                            process.createInstance(model);
```
```suggestion
                    ProcessInstance<M> pi = process.createInstance(model);
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/DataEventConsumer.java:
```diff
@@ -0,0 +1,61 @@
+    private Function<D, M> function;
+    private Class<D> dataEventClass;
+
+    public DataEventConsumer(Function<D, M> function,
+                             Class<D> dataEventClass) {
```
```suggestion
    public DataEventConsumer(Function<D, M> function, Class<D> dataEventClass) {
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/DataEventConsumer.java:
```diff
@@ -0,0 +1,61 @@
+            D eventData = EventUtils.readEvent(payload, dataEventClass);
+            M model = function.apply(eventData);
+            UnitOfWorkExecutor.executeInUnitOfWork(application.unitOfWorkManager(), () -> {
+                logger
+                    .debug(
```
```suggestion
                logger .debug(
```

cristianonicolai commented on jbpm/jbpm-flow/src/main/java/org/kie/kogito/event/impl/DefaultEventMarshaller.java:
```diff
@@ -0,0 +1,44 @@
+    private static final Logger logger = LoggerFactory.getLogger(DefaultEventMarshaller.class);
+
+    @Override
+    public <T, P extends AbstractProcessDataEvent<T>> String
+        marshall(T dataEvent, Function<T, P> cloudFunction, Optional<Boolean> isCloudEvent) {
```
```suggestion
    public <T, P extends AbstractProcessDataEvent<T>> String marshall(T dataEvent, Function<T, P> cloudFunction, Optional<Boolean> isCloudEvent) {
```

fjtirado replied:
Formatter changed to do not wrap method declaration

cristianonicolai commented on kogito-codegen/src/main/resources/class-templates/MessageProducerTemplate.java:
```diff
@@ -1,29 +1,21 @@
 
     Object emitter;
 
     Optional<Boolean> useCloudEvents = Optional.of(true);
```
@fjtirado this is resolved at compile time right, so why dont we inject a DefaultEventMarshaller or CloudEventMarshaller depending on that so we dont need to check at runtime? Instead we would have to concrete classes. Wouldnt that make things easier in the code we have to maintain?

fjtirado replied:
Im afraid this is not resolved at compilation time (I agree that it will be far easier if it was)
See generated code 
@org.eclipse.microprofile.config.inject.ConfigProperty(name = "kogito.messaging.as-cloudevents")
this is decided by configuration at runtime

cristianonicolai replied:
yep, you're right. Although I feel that is something that could be defined at compile time, but that's another story.

sutaakar reviewed (approved):
Looks good to me.

cristianonicolai reviewed (approved):
looks good, just some minor improvements could be done around formatting, like in here https://github.com/kiegroup/kogito-runtimes/pull/723/files#diff-266a17d98d95b932425d0fe0fa53e050R32

fjtirado commented:
> looks good, just some minor improvements could be done around formatting, like in here https://github.com/kiegroup/kogito-runtimes/pull/723/files#diff-266a17d98d95b932425d0fe0fa53e050R32

Swithed the formatter so it only wrap method parameters when stricly necessary

fjtirado commented:
jenkins retest this

fjtirado commented:
please jenkins retest this

fjtirado commented:
@kiegroup/gatekeepers  please merge this
